lib/util/pre_receive_util.js

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2016, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+/**
+ * This module contains the pre-recieve hook for a git-meta server.  Only
+ the meta-repo should use this hook.
+ */
+
+const NodeGit = require("nodegit");
+const GitUtil = require("./git_util");
+const split = require("split");
+const co = require("co");
+const path = require("path");
+const assert  = require("chai").assert;
+
+const NOTES_REF = "refs/notes/git-meta/subrepo-check";
+
+/**
+ * The identity function
+ */
+function identity(v) {
+    return v;
+}
+
+/**
+ * Map a function over an array, returning an array containing only
+ * the elements for which the function returns a truthy value.
+ *
+ * @return {Array}
+ */
+function flatmap(arr, fn) {
+    return arr.map(fn).filter(identity);
+}
+
+function *getParentTrees(parentEntries) {
+    const entryParentTreeEntries = parentEntries.filter(function isTree(entry) {
+        return entry.isTree();
+    });
+    return yield entryParentTreeEntries.map(
+        function *(entry) {
+            return yield entry.getTree();
+        });
+}
+
+const urlToLocalPath = Symbol();
+const checkSubmodule = Symbol();
+const checkTree = Symbol();
+const parentLoop = Symbol();
+class PreReceiveCheck {
```

Comment:
What methods do you need to mock and for what purpose? I don't see us providing a mock implementation of `PreReceive`. Actually, it might make sense to name this util something about its purpose, like `validate_meta_util` rather than its current caller.